swiftsimio/subset_writer.py

```diff
@@ -0,0 +1,172 @@
+    dataset_names : list of str
+        names of datasets found in the snapshot
+    """
+    if mask is not None:
+        for name in dataset_names:
+            if "Cells" not in name:
+                # get output dtype and size 
+                first_value = infile[name][0]
+                output_type = first_value.dtype
+                output_size = first_value.size
+                mask_size = compute_mask_size(mask, name)
+                if output_size != 1:
+                    output_shape = (mask_size, output_size)
+                else:
+                    output_shape = mask_size
+    
+                dataset_mask = get_mask_label(mask, name)
+                subset = read_ranges_from_file(infile[name], dataset_mask, output_shape = output_shape, output_type = output_type)
+                
+                # Write the subset
+                outfile.create_dataset(name, data=subset)
+
+def write_metadata(infile, outfile):
+    """
+    Copy over all the metadata from snapshot to output file
+
+    Parameters
+    ----------
+    infile : h5py.File
+        hdf5 file handle for input snapshot
+    outfile : h5py.File
+        hdf5 file handle for output snapshot
+    """
+    for field in infile.keys():
+        if not any([group_str in field for group_str in ["PartType"]]):
+            infile.copy(field, outfile)
+
+def find_datasets(input_file: h5py.File, dataset_names, path = None):
+    """
+    Recursively finds all the datasets in the snapshot and writes them to a list
+
+    Parameters
+    ----------
+    input_file : h5py.File
+        hdf5 file handle for snapshot
+    dataset_names : list of str
+        names of datasets found in the snapshot
+    path : str, optional
+        the path to the current location in the snapshot
+    """
+    if path != None:
+        keys = input_file[path].keys()
+    else:
+        keys = input_file.keys()
+        path = ""
+
+    for key in keys:
+        subpath = path + "/" + key
+        if isinstance(input_file[subpath], h5py.Dataset):
+            dataset_names.append(subpath)
+        elif input_file[subpath].keys() != None:
+            find_datasets(input_file, dataset_names, subpath)
+
+def write_subset(input_file: str, output_file: str, mask: SWIFTMask):
+    """
+    Writes subset of snapshot according to specified mask to new snapshot file
+
+    Parameters
+    ----------
+    input_file : str
+        path to input snapshot
+    output_file : str
+        path to output snapshot
+    mask : SWIFTMask
+        the mask used to define subset that is written to new snapshot
+    """
+    # Open the files
+    infile = h5py.File(input_file, "r")
+    outfile = h5py.File(output_file, "w")
+
+    # Find the datasets
+    dataset_names = []
+    find_datasets(infile, dataset_names)
+    
+    # Write metadata and data subset
+    write_metadata(infile, outfile)
+    write_datasubset(infile, outfile, mask, dataset_names)
```

Comment:
Can we be cheeky here and do `write_datsubset(infile,outfile,mask,find_datasets(infile))` with dataset_names a default parameter of `[]` in the function declaration above? I suppose this would require the addition of a return at the end of the for loop?